### ISSUE TITLE
Unable to expose Rust struct with `Vec` field

### ISSUE THREAD

**Issue by yonkeltron** [napi-rs/napi-rs#1752]
Hi, Thanks for making NAPI-RS!

I believe I've either found a bug in NAPI-RS or a bug in the documentation. In this case, I have written a Rust struct with a field with a `Vec` type signature. Confusingly, I am getting an error from NAPI-RS. Here is a reduced test case which is a complete `lib.rs` file in a fresh project created with `napi new`:

```rust
#![deny(clippy::all)]

#[macro_use]
extern crate napi_derive;

#[derive(Clone)]
#[napi]
pub struct Element {
  pub name: String,
}

#[napi]
pub struct Parent {
  pub elements: Vec<Element>,
}
```

And when I try to compile, I get this error:

```
error[E0277]: the trait bound `Vec<Element>: FromNapiValue` is not satisfied
  --> src/lib.rs:12:1
   |
12 | #[napi]
   | ^^^^^^^ the trait `FromNapiValue` is not implemented for `Vec<Element>`
   |
   = help: the trait `FromNapiValue` is implemented for `Vec<T>`
   = note: this error originates in the attribute macro `napi` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Now this is quite confusing to me since `Vec<Element>` should merely be the concrete type of `Vec<T>` which leads me to wonder if this is a bug. I did try to rewrite the struct like so:

```rust
#[napi]
pub struct Parent<T> {
  pub elements: Vec<T>,
}
```

But the error I got was even more confusing:

```
error[E0107]: missing generics for struct `Parent`
  --> src/lib.rs:13:12
   |
13 | pub struct Parent<T> {
   |            ^^^^^^ expected 1 generic argument
   |
note: struct defined here, with 1 generic parameter: `T`
  --> src/lib.rs:13:12
   |
13 | pub struct Parent<T> {
   |            ^^^^^^ -
help: add missing generic argument
   |
13 | pub struct Parent<T><T> {
   |                  +++
```

The second error message makes me wonder if this is a problem at the macro-level. Obviously, I'm not an expert.

Thanks so much for all your hard work on this project. NAPI-RS has tremendous potential and I look forward to using it even more.

Please let me know if I can provide more information!

Thanks,
+Jonathan

**Comment by lopo12123** [comment]
How about trying `#[napi(object)]`

look at here, https://napi.rs/docs/concepts/object, `#[napi]` means `class` while `#[napi(object)]` means `object`

``` rust
#![deny(clippy::all)]

#[macro_use]
extern crate napi_derive;

#[derive(Clone)]
#[napi(object)]  // change here
pub struct Element {
  pub name: String,
}

#[napi]  // you can keep it if it's a class
pub struct Parent {
  pub elements: Vec<Element>,
}
```

I copied the above code directly into my project, and it compiled correctly (the content outside the screenshot is my original project, don't worry about it)

![image](https://github.com/napi-rs/napi-rs/assets/59609929/f3158de2-2803-403d-a057-de42a10b5620)

---

If your original intention is that Element is a class, then this seems to be a mistake (I'm not sure).

I tried the following code but it failed to compile and I encountered the same error.

```rust

#[derive(Clone)]
#[napi(constructor)]  // change here
pub struct Element {
    pub name: String,
}

#[napi]
pub struct Parent {
    pub elements: Vec<Element>,
}
```

**Comment by bodymindarts** [comment]
+1 - I've run into this too!

**Comment by jssblck** [comment]
Running into this as well:

```rust
/// The result of the operation.
///
/// A function returning this type indicates that
/// it will attempt to continue through errors,
/// collecting them and returning them alongside
/// any successful results.
///
/// Such functions _should_ only result in an exception
/// due to extreme runtime issues, for example
/// if the process runs out of memory and times out
/// trying to allocate more.
#[derive(Clone, Eq, PartialEq, Debug)]
#[napi(object)]
pub struct Resilient<T, E> {
    pub ok: Option<T>,
    pub errors: Vec<Option<E>>,
}
```

Output snippet from `cargo build`:
```
error[E0107]: missing generics for struct `Resilient`
   --> src/lib.rs:151:12
    |
151 | pub struct Resilient<T, E> {
    |            ^^^^^^^^^ expected 2 generic arguments
    |
note: struct defined here, with 2 generic parameters: `T`, `E`
   --> src/lib.rs:151:12
    |
151 | pub struct Resilient<T, E> {
    |            ^^^^^^^^^ -  -
help: add missing generic arguments
    |
151 | pub struct Resilient<T, E><T, E> {
    |                     ++++++
```

**Comment by CPunisher** [comment]
When the fields of a struct are set to `pub`, napi will automatically generate the setter for that property, using `FromNapiValue` to convert the data passed from the JS side to Rust. Since all objects on the JS side are stored on the heap, we can actually only pass references to objects to the Rust side. This is why `Element` does not implement `FromNapiValue` while `&Element` does.

Therefore, you can set the fields to private and manually implement getters and setters based on references:

```rust
#[derive(Clone)]
#[napi(constructor)]
pub struct Element {
  pub name: String,
}

#[napi]
pub struct Parent {
  elements: Vec<Element>,
}

#[napi]
impl Parent {
  #[napi]
  pub fn set_elements(&mut self, elements: Vec<&Element>) {
    self.elements = elements.into_iter().map(|e| e.clone()).collect();
  }
}
```

You might think that this approach has a high communication overhead and that the data stored on the Rust side is a deep copy (modifications on one side do not affect the other, just like the approach described in https://napi.rs/docs/concepts/object). However, for data structures natively supported by napi, deep copies will also occur during communication between the two sides:
```rust
#[derive(Clone)]
#[napi(constructor)]
pub struct Element {
  pub names: Vec<String>,
}
```
```js
const e = new Element(["a"]);
console.log(e.name); // ['a']
e.name.push("b");
console.log(e.name); // ['a'] NO CHANGE
```

In short, data will be copied back and forth between the heap/stack on the Rust side and the heap on the JS side.
